backend/core/models/user.py

```diff
@@ -202,6 +202,13 @@ class User(AbstractUser):
                   " domains and the vouching mechanism.",
     )
 
+    settings = models.JSONField(
+        null=False,
+        blank=True,
+        default=dict,
+        help_text=_("The user' preferences.")
+    )
```

Comment:
What are the advantages of using a `JSONField` to store user settings? I haven't thought too much about this, but I feel like individual fields (maybe even in a separate model) would be easier to maintain. For example, when a new setting is required, its default value would be tracked automatically by Django migrations.

Reply:
Yes it's true that a `JSONField` forces us to write manual data migrations if we want to set a default value for a new setting. With one model field per setting it's automatic.

Nevertheless I considered the `JSONField` because I think it can help us to easily solve several "complex" problems.

The main problem: some settings can be poll specific, and not all polls have the same settings. If I create a single table to store the settings, let's say with a poll foreign key to not force every poll to behave the same way, how do we store settings that are specific to only the `videos` and not to the other? (I think this problem has a precise name, is it polymorphic table?)

Let's imagine a setting `recommendations__max_duration` that defines the maximum duration of videos displayed on the recommendations page. I see two ways to add this setting:

**(a)** if I add a column `recommendations__max_duration` in the dedicated settings table, all records will have this column, even those that don't have this concept of max duration. The value of this column will probably be null for those records, and the table will grow "horizontally" each time we add a new setting. I personally don't like tables like this because it implies that records can store values that have no meaning for them ; and because the records don't represent a clear concept anymore, but rather an aggregation or possibly unrelated ones.

**(b)** I could add instead a new column named `specific` to store a JSON of all settings exclusively related to the poll. This hybrid solution has some settings stored in columns with regular data type and some settings stored in a `JSONField`. It's better than the solution **(a)** but I think it adds extra work without bringing more comfort: we need to do SQL joins to get the settings when we have a user instance, and depending on the poll we need to use a specific serializer to read the settings.

The advantage of a single `JSONField` :
- less code
- no SQL join is required, the settings are available each time a user instance is available
- the polls can have specific settings, adding new one doesn't require to update a table shape
- some settings can be poll agnostic, like `email_notification__platform_updates`

And the main disadvantage is that we will have to manually manage the default values of settings.

Reply:
Ok, if we need to store settings that are specific to a given poll, a JSONField is indeed the best way.